src/actions/jira/jira_client.ts

```diff
@@ -0,0 +1,223 @@
+import * as querystring from "querystring"
+import * as https from "request-promise-native"
+import * as URL from "url"
+
+import { Document } from "adf-builder"
+
+export interface Credentials {
+  access_token: string
+  expires_in: number
+  refresh_token?: string
+  scope: string
+  token_type: string
+}
+
+interface JiraIssue {
+  project: {
+    id: string,
+  }
+  summary?: string
+  description?: string
+  url?: string
+  issuetype: {
+    id: string,
+  }
+}
+
+export class JiraClient {
+  apiVersion: string
+  redirectUri: string
+  tokens?: Credentials
+  cloudId?: string
+
+  constructor(redirectUri: string, tokens?: Credentials, apiVersion = "3") {
+    this.redirectUri = redirectUri
+    this.tokens = tokens
+    this.apiVersion = apiVersion
+  }
+
+  generateAuthUrl(encryptedState: string, scope: string) {
+    const url = new URL.URL("https://auth.atlassian.com/authorize")
+    url.search = querystring.stringify({
+      audience: "api.atlassian.com",
+      client_id: process.env.JIRA_CLIENT_ID,
+      scope,
+      redirect_uri: this.redirectUri,
+      state: encryptedState,
+      response_type: "code",
+      prompt: "consent",
+    })
+    return url.toString()
+  }
+
+  async setCloudIdFromTokens() {
+    if (!this.tokens) {
+      throw "unauthenticated"
+    }
+    this.cloudId = await this.getCloudIdFromTokens()
+    return this.cloudId
+  }
+
+  async getCloudIdFromTokens() {
+    if (!this.tokens) {
+      throw "unauthenticated"
+    }
+    const options = {
+      url: "https://api.atlassian.com/oauth/token/accessible-resources",
+      headers: {
+        Authorization: `Bearer ${this.tokens.access_token}`,
+      },
+      json: true,
+    }
+
+    const response = await https.get(options).promise()
+    if (response.length === 0) {
+      throw "no cloudId"
+    } else {
+      return response[0].id
+    }
+  }
+
+  async getToken(code: string) {
+    let tokens: Credentials = await https.post({
+      url: "https://auth.atlassian.com/oauth/token",
+      headers: {
+        "Content-Type": "application/json",
+      },
+      body: {
+        grant_type: "authorization_code",
+        code,
+        client_id: process.env.JIRA_CLIENT_ID,
+        client_secret: process.env.JIRA_CLIENT_SECRET,
+        redirect_uri: this.redirectUri,
+      },
+      json: true,
+    }).promise()
+    if (tokens.refresh_token) {
+      tokens = await this.getRefreshToken(tokens.refresh_token)
+    }
+    return tokens
+  }
+
+  async getRefreshToken(refresh: string) {
```

Comment:
We're storing the refresh token, but we're not using it to request a new access token once the access token expires after 60 minutes. Can we add this functionality in on form load?